### PR TITLE
First approach of rule SA1140 MaximumLineLength

### PR DESCRIPTION
**Pull request by pkess**
Hi,
i just created a rule to meet #1808.
The rule will not raise any warnings or errors unless a maximum line length is configured in stylecop.json

Please leave comments

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/ReadabilityResources.resx [hunk 1]
```diff
@@ -534,4 +534,13 @@
   <data name="SX1101Title" xml:space="preserve">
     <value>Do not prefix local calls with 'this.'</value>
   </data>
+  <data name="SA1140Description" xml:space="preserve">
+    <value>Code should be wrapped to a maximum line length. See stylecop.settings</value>
```
Comment (bjornhellander) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/ReadabilityResources.resx]: The configuration file is called stylecop.json

Comment (bjornhellander) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/ReadabilityResources.resx]: The configuration file is called stylecop.json

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1140MaximumLineLength.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    using System;
+    using System.Collections.Generic;
+    using System.Collections.Immutable;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.Diagnostics;
+    using StyleCop.Analyzers.Helpers;
+
+    /// <summary>
+    /// The C# comment does not contain any comment text.
```
Comment (bjornhellander) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1140MaximumLineLength.cs]: Copy paste

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1140MaximumLineLength.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    /// </summary>
+    /// <remarks>
+    /// <para>A violation of this rule occurs whenever the code contains a C# comment which does not contain any
```
Comment (bjornhellander) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1140MaximumLineLength.cs]: Copy paste

File: documentation/SA1140.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+:memo: This rule is new for StyleCop Analyzers, and was not present in StyleCop Classic.
+
+## Cause
+
+A maximum number of characters in one line can lead to a better readability on screen and printed code.
+
+## Rule description
+
+This rule is disabled by default. To enabel add the setting for maximum line length to your stylecop.settings.
```
Comment (bjornhellander) [inline documentation/SA1140.md]: enabel -> enable

Comment (bjornhellander) [inline documentation/SA1140.md]: enabel -> enable
The configuration file is called stylecop.json

File: documentation/SA1140.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+## How to fix violations
+
+To fix a violation of this rule, use the appropriate literal suffix instead of casting the value in code.
```
Comment (bjornhellander) [inline documentation/SA1140.md]: Copy paste

**Comment by pkess** [comment]
Corrected some copy paste errors and added example for documentation

**Comment by pkess** [comment]
Any news here?

**Comment by vweijsters** [comment]
I'll have a look this week. Marked it 'do not merge' to signify that the discussion in #1808 has not come to a conclusion yet.

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1140UnitTests.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+            this.SetLineLengthSettings(34);
+            await this.VerifyCSharpDiagnosticAsync(testCode, new[] { diagnosticLine1, diagnosticLine2, diagnosticLine3, diagnosticLine4 }, CancellationToken.None).ConfigureAwait(false);
+
+            this.SetLineLengthSettings(35);
+            await this.VerifyCSharpDiagnosticAsync(testCode, new[] { diagnosticLine2, diagnosticLine3, diagnosticLine4 }, CancellationToken.None).ConfigureAwait(false);
+
+            this.SetLineLengthSettings(41);
+            await this.VerifyCSharpDiagnosticAsync(testCode, new[] { diagnosticLine4 }, CancellationToken.None).ConfigureAwait(false);
+        }
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/ReadabilityRules/SA1140UnitTests.cs]: This is missing a test case where all lines are shorter than the maximum line length

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1140MaximumLineLength.cs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+    /// <summary>
+    /// Maximum line length
+    /// </summary>
+    /// <remarks>
+    /// <para>A violation of this rule occurs whenever the code contains lines with more characters
+    /// than allowed in stylecop.json</para>
+    /// </remarks>
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal class SA1140MaximumLineLength : DiagnosticAnalyzer
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1140MaximumLineLength.cs]: This should be marked with `[NoCodeFix]`

File: StyleCop.Analyzers/StyleCop.Analyzers/Settings/ObjectModel/ReadabilitySettings.cs [hunk 8]
```diff
@@ -9,11 +9,21 @@ namespace StyleCop.Analyzers.Settings.ObjectModel
...
         /// </summary>
         [JsonConstructor]
         protected internal ReadabilitySettings()
         {
+            this.lineLength = 0;
         }
+
+        public int LineLength =>
+            this.lineLength;
     }
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/Settings/ObjectModel/ReadabilitySettings.cs]: LineLength should be named MaximumLineLength to provide more clarity to someone reading the `stylecop.json` file.

File: documentation/ReadabilityRules.md [hunk 9]
```diff
@@ -39,3 +39,4 @@ Identifier | Name | Description
 [SA1134](SA1134.md) | AttributesMustNotShareLine | An attribute is placed on the same line of code as another attribute or element. 
 [SA1136](SA1136.md) | EnumValuesShouldBeOnSeparateLines | Multiple enum values are placed on the same line of code. 
 [SA1139](SA1139.md) | UseLiteralsSuffixNotationInsteadOfCasting | Use literal suffix notation instead of casting. 
+[SA1140](SA1140.md) | SA1140MaximumLineLength | Code should be wrapped to a maximum line length.
```
Comment (vweijsters) [inline documentation/ReadabilityRules.md]: `SA1140MaximumLineLength` should be `MaximumLineLength`, to be inline with the other entries

File: documentation/SA1140.md [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+  <td>SA1140</td>
+</tr>
+<tr>
+  <td>Category</td>
+  <td>Readability Rules</td>
+</tr>
+</table>
+
+:memo: This rule is new for StyleCop Analyzers, and was not present in StyleCop Classic.
+
```
Comment (vweijsters) [inline documentation/SA1140.md]: Please add a remark that this was added in 1.1.0

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1140MaximumLineLength.cs [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+        /// </summary>
+        public const string DiagnosticId = "SA1140";
+        private static readonly LocalizableString Title = new LocalizableResourceString(nameof(ReadabilityResources.SA1140Title), ReadabilityResources.ResourceManager, typeof(ReadabilityResources));
+        private static readonly LocalizableString MessageFormat = new LocalizableResourceString(nameof(ReadabilityResources.SA1140MessageFormat), ReadabilityResources.ResourceManager, typeof(ReadabilityResources));
+        private static readonly LocalizableString Description = new LocalizableResourceString(nameof(ReadabilityResources.SA1140Description), ReadabilityResources.ResourceManager, typeof(ReadabilityResources));
+        private static readonly string HelpLink = "https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1140.md";
+
+        private static readonly DiagnosticDescriptor Descriptor =
+            new DiagnosticDescriptor(DiagnosticId, Title, MessageFormat, AnalyzerCategory.ReadabilityRules, DiagnosticSeverity.Warning, AnalyzerConstants.EnabledByDefault, Description, HelpLink);
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1140MaximumLineLength.cs]: This should have `AnalyzerConstants.DisabledByDefault`

Comment (pkess) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1140MaximumLineLength.cs]: Am i right that the `AnalyzerConstants.DisabledByDefault` will set the Severity Level to None by default?
I would like this rule to check the line length as soon as a user defined a maximum in his config. Thats the way it works at the moment.

File: documentation/ReadabilityRules.md (hunk 9 above)
Comment (pkess) [inline documentation/ReadabilityRules.md]: Done

File: documentation/SA1140.md (hunk 10 above)
Comment (pkess) [inline documentation/SA1140.md]: Done

**Comment by pkess** [comment]
Ok, i think finally i implemented all of your requests exept the `DisabledByDefault` constant.
I am sorry for all those commits but i did not had visual studio installed on the pc i used to implement all those changes.

Please give feedback

**Comment by pkess** [comment]
Hey guys,

Did you already checkout my last changes?

**Comment by cristobalito** [comment]
*apologies for unsolicited commenting on this PR - I've nothing to do with project but was looking for this sort of functionality as a user*

@pkess - nice work. It would be great to see this merged into the project. While i like your approach w.r.t. to the outstanding issue on this PR, esp. for its convenience, it strikes me as a non-standard way to configure the rule. As a user, I would expect that the value specified in my ruleset ultimately governs whether or not the rule is on. One major benefit of this, is that there is only one obvious place I have to go in order to disable the rule, should I so desire.

In order to break the impasse here, would you be willing to compromise here to get the majority of the benefit into the codebase?

**Comment by pkess** [comment]
Hi,

Thank you for your comment.
In fact I think the way of activating the rule by specification of a line length is very clear. There is another way to specific set the severity level of rules. I don't like to implement a separate way of activation.

Additionally I don't have a machine with Windows and visual studio running now (and will not have any in the next month's) so I am not able to fix it quite now. Feel free to fork my branch and implement it as you like. I would be excited!

**Comment by sharwell** [comment]
@pkess I'm won't hide that I'm not a fan of this rule, but after I get some of the other blocking issues (e.g. C# 7) taken care of I will take a serious look at this PR and see if we can get it included as a disabled-by-default rule.

**Comment by LordBenjamin** [comment]
@sharwell - is there any timeline for including this functionality? It would be useful to know before we decide whether to reference another third-party analyzer or write our own equivalent.

**Comment by sharwell** [comment]
@LordBenjamin Not yet. Feel free to keep reminding me.

**Comment by cristobalito** [comment]
@sharwell If this rule wasn't enabled by default (as per original PR), would you be happy to merge it? If so, I'd be happy to take a look at doing that tonight.

**Comment by sharwell** [comment]
@cristobalito I haven't had a chance to do the code review yet. If the only issues are the merge conflicts and changing it to disabled by default, then those are items we can quickly clean up when we get ready to merge. Thank you though. 😄

**Comment by promontis** [comment]
@sharwell is it possible to merge this?

**Comment by PeterRockstars** [comment]
Any progress on this? Would love to see this rule implemented and it seems that we are on the home stretch. Anyone care to take us the last few yards?

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/ReadabilityResources.resx [hunk 12]
```diff
@@ -534,4 +534,13 @@
   <data name="SX1101Title" xml:space="preserve">
     <value>Do not prefix local calls with 'this.'</value>
   </data>
+  <data name="SA1140Description" xml:space="preserve">
+    <value>Code should be wrapped to a maximum line length. See stylecop.json</value>
+  </data>
+  <data name="SA1140MessageFormat" xml:space="preserve">
+    <value>Maximum line length exceeded</value>
```
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/ReadabilityResources.resx]: I think it would be good to change this to "Maximum line length of {0} exceeded" and make it include the set line length.

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1140MaximumLineLength.cs [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+            // Only proceed if rule is enabled
+            if (readabilityRules.MaximumLineLength > 0)
+            {
+                SyntaxNode root = context.Tree.GetCompilationUnitRoot(context.CancellationToken);
+                var fileText = context.Tree.GetText(context.CancellationToken);
+                foreach (var line in fileText.Lines)
+                {
+                    if ((line.End - line.Start) > readabilityRules.MaximumLineLength)
+                    {
+                        var location = root.SyntaxTree.GetLocation(line.Span);
```
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1140MaximumLineLength.cs]: Maybe only report the diagnostic on the bit that is too long so the user knows how much the line is too long.

File: documentation/SA1140.md [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+	}
+  }
+}
+```
+
+A value of 0 (default) for `maximumLineLength` disables the rule.
+
+## Known issues
+
+Tabs in source file are not expanded to spaces for evaluation of the line length. Only use this rule if spaces are used for indention.
```
Comment (pdelvo) [inline documentation/SA1140.md]: Should we do something about this? We could count the number of tabs in the line and use the configured tab width for that

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1140MaximumLineLength.cs [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+    using System;
+    using System.Collections.Generic;
+    using System.Collections.Immutable;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.Diagnostics;
+    using StyleCop.Analyzers.Helpers;
+
+    /// <summary>
+    /// Maximum line length
```
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1140MaximumLineLength.cs]: All(?) other rules contain the documentation as part of the xml documentation

File: documentation/SA1140.md (hunk 14 above)
Comment (sharwell) [reply documentation/SA1140.md]: At the time this was first implemented, tab width might not have been included in stylecop.json. It's now available there so the feature could be implemented in this rule too.

**Comment by danpetitt** [comment]
Any news on this? My new fellow devs (company merged) prefer to operate a max line length style and I want something to pre-emptively warn me

**Comment by insulind** [comment]
Would be very happy to see this as a feature. Any news on progress?

**Comment by admsugar** [comment]
Any update on this?

**Comment by jnm2** [comment]
I believe all updates will be seen in this thread or the proposal issue, if that's what you're asking.

**Comment by jhoy1020** [comment]
Is this change going to be checked in?  Trying to get my team to standardize on a max line length is impossible without some kind of preventative check. Per documentation, it is recommended to have a max line length of 120 https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions. It would be really helpful to have this change to help enforce this rule.
